### PR TITLE
[imgui] Fix find dependencies

### LINKED ISSUE #14753: [imgui] issue with opengl bindings
****

**Is your feature request related to a problem? Please describe.**
`vcpkg install imgui[opengl3-glbinding-binding,glfw-binding]:x64-windows` will build with glew.

resulting
```
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewActiveTexture referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBlendEquation referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBlendFuncSeparate referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBindBuffer referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBufferData referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteBuffers referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGenBuffers referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewAttachShader referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBlendEquationSeparate referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewCompileShader referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewCreateProgram referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewCreateShader referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteProgram referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteShader referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDetachShader referenced in function "void __cdecl ImGui_ImplOpenGL3_DestroyDeviceObjects(void)" (?ImGui_ImplOpenGL3_DestroyDeviceObjects@@YAXXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewEnableVertexAttribArray referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetAttribLocation referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetProgramInfoLog referenced in function "bool __cdecl CheckProgram(unsigned int,char const *)" (?CheckProgram@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetProgramiv referenced in function "bool __cdecl CheckProgram(unsigned int,char const *)" (?CheckProgram@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetShaderInfoLog referenced in function "bool __cdecl CheckShader(unsigned int,char const *)" (?CheckShader@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetShaderiv referenced in function "bool __cdecl CheckShader(unsigned int,char const *)" (?CheckShader@@YA_NIPEBD@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGetUniformLocation referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewLinkProgram referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewShaderSource referenced in function "bool __cdecl ImGui_ImplOpenGL3_CreateDeviceObjects(void)" (?ImGui_ImplOpenGL3_CreateDeviceObjects@@YA_NXZ) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewUniform1i referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewUniformMatrix4fv referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewUseProgram referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewVertexAttribPointer referenced in function "void __cdecl ImGui_ImplOpenGL3_SetupRenderState(struct ImDrawData *,int,int,unsigned int)" (?ImGui_ImplOpenGL3_SetupRenderState@@YAXPEAUImDrawData@@HHI@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDrawElementsBaseVertex referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBindSampler referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewBindVertexArray referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewDeleteVertexArrays referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
[build] imguid.lib(imgui_impl_opengl3.cpp.obj) : error LNK2019: unresolved external symbol __imp___glewGenVertexArrays referenced in function "void __cdecl ImGui_ImplOpenGL3_RenderDrawData(struct ImDrawData *)" (?ImGui_ImplOpenGL3_RenderDrawData@@YAXPEAUImDrawData@@@Z) [****.vcxproj]
```

**Proposed solution**

So lets make the problem clear first. For opengl there are two types of helper library commonly used.

1.  OpenGL symbol loader: e.g. glbinding, glew, gl3w, etc.
2. OpenGL window and context helper: e.g. glfw, glut

In this case, imgui[opengl3-glbinding-binding,glfw-binding] is actually built as imgui[opengl3-glew-binding,glfw-binding], resulting glew symbols in the `imguid.lib` file.

This should be fix.

**Describe alternatives you've considered**

Even the above issue is fixed, there will be potential problem:
1. project A use glew +  glfw
2. project B use glbinding + glfw
3. project C use DX12
4. ...

since there will be no way to let one installation of vcpkg to supply polymorphic `imgui.lib`, this problem will never be solved if vcpkg building bindings into library code.

The only solution I can come up with is copying bindings `imgui_impl_*.cpp` files into imgui's `include` folder. Let the use do

```
// imgui_impl.cpp file
// assuming the user use glbinding + glfw

// user do necessary definition

#include <bindings/imgui_impl_glfw.cpp>
#include <bindings/imgui_impl_opengl3.cpp>
```
and let the user compile this source and link the obj file to its executable.

### PR DESCRIPTION
**Pull request by JackBoosY**
- Add dependent `find_dependency`.

Fixes #14753.

### CODE REVIEW

**Comment by JackBoosY** [comment]
Special thanks to @thedmd for giving me technical help.

**Comment by JackBoosY** [comment]
@cloudhan Can you please test this PR?

Thanks.

**Comment by JackBoosY** [comment]
The sample solution provided by @cloudhan has been successfully tested.

**Review by PhoebeHui — APPROVED** [review]

File: ports/imgui/CMakeLists.txt [hunk 1]
```diff
@@ -6,135 +6,225 @@ set(CMAKE_DEBUG_POSTFIX d)
...
 endif()
 
-if(IMGUI_BUILD_OPENGL3_GLAD_BINDING)
-	find_package(glad REQUIRED)
-	get_target_property(GLAD_INCLUDE_DIRS glad::glad INTERFACE_INCLUDE_DIRECTORIES)
-	target_include_directories(${PROJECT_NAME} PRIVATE ${GLAD_INCLUDE_DIRS})
-	target_sources(${PROJECT_NAME} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/examples/imgui_impl_opengl3.cpp)
+if(IMGUI_INSTALL_OPENGL3_GLAD_BINDING)
+    find_package(glad REQUIRED)
+    add_library(imgui_impl_opengl3_glda INTERFACE)
```
Comment (mwthinker) [inline ports/imgui/CMakeLists.txt:133]: **Wrong:** imgui_impl_opengl3_glda
**should be:** imgui_impl_opengl3_glad

File: ports/imgui/CMakeLists.txt [hunk 2]
```diff
@@ -6,135 +6,225 @@ set(CMAKE_DEBUG_POSTFIX d)
...
+    add_dependencies(imgui_impl_opengl3_glda ${PROJECT_NAME})
+    set_property(TARGET imgui_impl_opengl3_glda PROPERTY INTERFACE_SOURCES $<INSTALL_INTERFACE: examples/imgui_impl_opengl3.cpp>)
+    set_property(TARGET imgui_impl_opengl3_glda PROPERTY INTERFACE_LINK_LIBRARIES glad::glad ${PROJECT_NAME})
+    set_property(TARGET imgui_impl_opengl3_glda PROPERTY INTERFACE_COMPILE_DEFINITIONS IMGUI_IMPL_OPENGL_LOADER_GLAD)
+    list(APPEND BINDING_TARGETS imgui_impl_opengl3_glda)
```
Comment (JackBoosY) [inline ports/imgui/CMakeLists.txt:138]:
```suggestion
    add_library(imgui_impl_opengl3_glad INTERFACE)
    add_dependencies(imgui_impl_opengl3_glad ${PROJECT_NAME})
    set_property(TARGET imgui_impl_opengl3_glad PROPERTY INTERFACE_SOURCES $<INSTALL_INTERFACE: examples/imgui_impl_opengl3.cpp>)
    set_property(TARGET imgui_impl_opengl3_glad PROPERTY INTERFACE_LINK_LIBRARIES glad::glad ${PROJECT_NAME})
    set_property(TARGET imgui_impl_opengl3_glad PROPERTY INTERFACE_COMPILE_DEFINITIONS IMGUI_IMPL_OPENGL_LOADER_GLAD)
    list(APPEND BINDING_TARGETS imgui_impl_opengl3_glad)
```

File: ports/imgui/CMakeLists.txt (hunk 1 above)
Comment (JackBoosY) [reply ports/imgui/CMakeLists.txt:133]: Typo, thanks!

File: ports/imgui/CMakeLists.txt [hunk 3]
```diff
@@ -6,135 +6,225 @@ set(CMAKE_DEBUG_POSTFIX d)
...
+set(BINDING_TARGETS )
+set(BINDINGS_SOURCES)
+if(IMGUI_INSTALL_ALLEGRO5_BINDING)
+    find_path(ALLEGRO5_INCLUDE_DIRS allegro5/allegro.h)
+    add_library(imgui_impl_allegro5 INTERFACE)
+    set_property(TARGET imgui_impl_allegro5 PROPERTY INTERFACE_SOURCES $<INSTALL_INTERFACE: examples/imgui_impl_allegro5.cpp>)
+    set_property(TARGET imgui_impl_allegro5 PROPERTY INTERFACE_INCLUDE_DIRECTORIES ${ALLEGRO5_INCLUDE_DIRS})
+    set_property(TARGET imgui_impl_allegro5 PROPERTY INTERFACE_LINK_LIBRARIES ${PROJECT_NAME})
+    list(APPEND BINDING_TARGETS imgui_impl_allegro5)
+    list(APPEND BINDINGS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/examples/imgui_impl_allegro5.cpp)
```
Comment (BillyONeal) [inline ports/imgui/CMakeLists.txt:41]: It's strange that before we were building these source files into the target but now we're just copying them into the examples directory. It seems like that could be a breaking change?

File: ports/imgui/CMakeLists.txt [hunk 4]
```diff
@@ -6,135 +6,225 @@ set(CMAKE_DEBUG_POSTFIX d)
...
-if(IMGUI_BUILD_ALLEGRO5_BINDING)
-	find_path(ALLEGRO5_INCLUDE_DIRS allegro5/allegro.h)
-	target_include_directories(${PROJECT_NAME} PRIVATE ${ALLEGRO5_INCLUDE_DIRS})
-	target_sources(${PROJECT_NAME} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/examples/imgui_impl_allegro5.cpp)
+set(BINDING_TARGETS )
+set(BINDINGS_SOURCES)
+if(IMGUI_INSTALL_ALLEGRO5_BINDING)
+    find_path(ALLEGRO5_INCLUDE_DIRS allegro5/allegro.h)
+    add_library(imgui_impl_allegro5 INTERFACE)
+    set_property(TARGET imgui_impl_allegro5 PROPERTY INTERFACE_SOURCES $<INSTALL_INTERFACE: examples/imgui_impl_allegro5.cpp>)
```
Comment (BillyONeal) [inline ports/imgui/CMakeLists.txt:37]:
```suggestion
    set_property(TARGET imgui_impl_allegro5 PROPERTY INTERFACE_SOURCES $<INSTALL_INTERFACE:examples/imgui_impl_allegro5.cpp>)
```

That extra space shouldn't be in any of these.

File: ports/imgui/CMakeLists.txt [hunk 5]
```diff
@@ -6,135 +6,225 @@ set(CMAKE_DEBUG_POSTFIX d)
...
+    add_library(imgui_impl_opengl2 INTERFACE)
+    add_dependencies(imgui_impl_opengl2 ${PROJECT_NAME})
+    set_property(TARGET imgui_impl_opengl2 PROPERTY INTERFACE_SOURCES $<INSTALL_INTERFACE: examples/imgui_impl_opengl2.cpp>)
+    set_property(TARGET imgui_impl_opengl2 PROPERTY INTERFACE_LINK_LIBRARIES Opengl32 ${PROJECT_NAME})
+    list(APPEND BINDING_TARGETS imgui_impl_opengl2)
+    list(APPEND BINDINGS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/examples/imgui_impl_opengl2.cpp)
 endif()
 
-if(IMGUI_BUILD_OPENGL3_GLEW_BINDING)
-	find_package(glew REQUIRED)
```
Comment (BillyONeal) [inline ports/imgui/CMakeLists.txt:76]: It seems like fixing the actual bug here could be done with
1. Add `target_link_libraries(${PROJECT_NAME} PUBLIC GLEW::GLEW)` 
2. Remove of `target_include_directories(${PROJECT_NAME} PRIVATE ${GLEW_INCLUDE_DIRS})`

rather than needing to duplicate all these `add_library` etc. calls.

File: ports/imgui/CMakeLists.txt [hunk 6]
```diff
@@ -6,135 +6,225 @@ set(CMAKE_DEBUG_POSTFIX d)
...
+    set_property(TARGET imgui_impl_glut PROPERTY INTERFACE_SOURCES $<INSTALL_INTERFACE: examples/imgui_impl_glut.cpp>)
+    set_property(TARGET imgui_impl_glut PROPERTY INTERFACE_LINK_LIBRARIES GLUT::GLUT ${PROJECT_NAME})
+    list(APPEND BINDING_TARGETS imgui_impl_glut)
+    list(APPEND BINDINGS_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/examples/imgui_impl_glut.cpp)
 endif()
 
-if(IMGUI_BUILD_METAL_BINDING)
-	target_sources(${PROJECT_NAME} PRIVATE ${CMAKE_CURRENT_SOURCE_DIR}/examples/imgui_impl_metal.mm)
+if(IMGUI_INSTALL_METAL_BINDING)
+    add_library(imgui_impl_metal INTERFACE)
```
Comment (BillyONeal) [inline ports/imgui/CMakeLists.txt:103]: Creating targets for these bits that might get copied into a user program doesn't seem related to the bug you're fixing. Is there a customer asking for this?

**Comment by ras0219-msft** [comment]
This PR makes very large, breaking changes to how users consume imgui through vcpkg; it moves the example backends out of the main library file and injects the sources into all consuming DLLs, static libs, and executables.

I don't think this large change is warranted by the actual issue (https://github.com/microsoft/vcpkg/issues/14753).

The linked PR in imgui has been in development for the last two and a half years, so I don't think it's appropriate to assume it will be merged very soon.

File: ports/imgui/CMakeLists.txt (hunk 3 above)
Comment (JackBoosY) [reply ports/imgui/CMakeLists.txt:41]: Yes, in fact, imgui is officially marked as a header-only library, and these source files should be compiled at the same time as the private project instead of being used as a library.

Comment (JackBoosY) [reply ports/imgui/CMakeLists.txt:41]: See [official reply](https://github.com/ocornut/imgui/pull/1713#issuecomment-378858117).

**Comment by JackBoosY** [comment]
@ras0219-msft The changes to this PR come from part of https://github.com/ocornut/imgui/pull/1713.

I think it is inappropriate for us to force imgui to be used as a library, which is also contrary to the official intention. This caused the previous problem that the variables in imgui were set twice(issue https://github.com/microsoft/vcpkg/issues/9660).

The modification in this PR fits [the official definition](https://github.com/ocornut/imgui#usage) of imgui (as a header-only library), and many people believe that this change is effective and reasonable. You can check the user's response in the upstream PR I provided.

**Comment by xarthurx** [comment]
@strega-nil

**Comment by xarthurx** [comment]
@strega-nil May I ask when this will be reviewed and merged?

**Comment by BillyONeal** [comment]
@xarthurx

This PR is controversial because:

1. It does something that upstream appears to be explicitly rejecting right now
2. It's a potentially breaking change for existing imgui customers

(See Robert's reply https://github.com/microsoft/vcpkg/pull/15063#issuecomment-747754005 for more details)

so it doesn't look like there's support to merge this right now, so it isn't a timeframe thing. Jack and Robert presumably need to have further discussion to figure out what happens.

**Comment by ras0219** [comment]
From the imgui README:
>The core of Dear ImGui is self-contained within a few platform-agnostic files which you can easily compile in your application/engine. They are all the files in the root folder of the repository (imgui.cpp, imgui.h, imgui_demo.cpp, imgui_draw.cpp etc.).
>
>No specific build process is required. You can add the .cpp files to your existing project.

Imgui is _not_ a header-only library; there are several files that must be compiled in order to use it. The official instructions for the core code are (almost perfectly) equivalent to linking it as a static library.

There are three differences between prebuilding it as a static library compared to injecting the code into the consumer project:
1. Injecting the code causes it to inherit macros and compiler settings from the consumer
2. Injecting the code makes it _extremely_ restrictive to write additional libraries on top of imgui. It forces these other libraries to also inject their code into the consumer to maintain the same contract (inherit compiler settings).
3. It forces the final consumer to build the entire transitive application from source instead of being able to use binary caching.

The real reason why `imgui` recommends adding the sources to your code and discourages a separate static library build is because:

>[Building separately] is a little harmful as it prevents people from configuring ImGui for their application via imconfig.h.

Modifying the sources of imgui is expected to be a way to configure the library. However, this change (injecting the source into the consumer) does not fundamentally fix this issue without further efforts to ensure that a user-provided `imconfig.h` would be selected over the file in `include/`.

Today, the user can overlay a modified version of the `imgui` port that copies a private `imconfig.h` over the sources before building; something like:
```cmake
file(COPY "${CMAKE_CURRENT_LIST_DIR}/imconfig.h" DESTINATION "${SOURCE_PATH}")
```
which will work perfectly with building a static library and still enables full binary caching of `imgui` and any libraries that depend on it (such as `imgui-sfml`).

The primary issue in #9660 would have been made no better by this change; it's net neutral (or worse!). The problem was that the `imgui-sfml` library was being built as a DLL and thus having a separate copy of all the imgui global variables as the user's final application. The only correct fix for this problem was to build `imgui-sfml` statically, which we've done. Furthermore, this change actually introduces an expectation violation because if the user _did_ successfully leverage these changes to configure imgui by replacing `imconfig.h` in the user's project, the sources built as part of `imgui-sfml` would not have that configuration and possibly re-introduce a similar bug!

Finally, the linked PR (ocornut/imgui#1713) has been in limbo for several years; it is unclear if it will be adopted as an official build. Even if it was, the author explicitly stated that the only purpose they see in having this build is to maintain the example apps:

>... to provide the example applications I will maintain premake/cmake ....

which means there would still be no official buildsystem for consuming it as a library (since the author intends the user to copy the source code into their project).


Even if I have committed several egregious errors above, we should separate fixing the issue in question (https://github.com/microsoft/vcpkg/issues/14753, which really just needs some `find_dependency()` & friends) with fundamentally shifting the way we currently provide imgui.

**Comment by JackBoosY** [comment]
@ras0219 Fine, I will separate the new compilation method of bindings to another PR.

File: ports/imgui/CMakeLists.txt (hunk 5 above)
Comment (JackBoosY) [reply ports/imgui/CMakeLists.txt:76]: Done.

**Comment by JackBoosY** [comment]
Test complete locally on `x64-windows` / `x64-windows-static`.

**Review by ras0219 — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for your edits!
